### PR TITLE
MAMORU-813 Collect Daemons statistics

### PR DESCRIPTION
**Pull request by gofmanaa**
Feat/sniffer statistics

### CODE REVIEW

File: mamoru-core/src/blockchain_data/mod.rs [hunk 1]
```diff
@@ -119,6 +121,16 @@ impl<T: BlockchainCtx> BlockchainDataBuilder<T> {
         self.block = Some((block_id.into(), block_hash.into()));
     }
 
+    pub fn set_statistic(&mut self, blocks: u64, transactions: u64, events: u64, call_traces: u64) {
+        self.statistic = Some(Statistic {
+            source: self.source,
```
Comment (sokyrko) [inline mamoru-core/src/blockchain_data/mod.rs:126]: It looks like there is no reason to duplicate `source` field, remove it.

File: mamoru-core/src/blockchain_data/mod.rs [hunk 2]
```diff
@@ -187,6 +223,10 @@ impl<T> BlockchainData<T> {
         self.source
     }
 
+    pub fn statistic(&self) -> &Statistic {
```
Comment (sokyrko) [inline mamoru-core/src/blockchain_data/mod.rs:226]: `statistics`

File: mamoru-core/src/blockchain_data/mod.rs [hunk 3]
```diff
@@ -153,11 +166,33 @@ impl<T: BlockchainCtx> BlockchainDataBuilder<T> {
...
+#[derive(Clone, Debug)]
+pub struct Statistic {
+    pub source: DataSource,
+    pub blocks: u64,
+    pub transactions: u64,
+    pub events: u64,
+    pub call_traces: u64,
+}
+
+impl Default for Statistic {
```
Comment (sokyrko) [inline mamoru-core/src/blockchain_data/mod.rs:178]: Use `#[derive(Default)]` instead.

File: mamoru-core/src/blockchain_data/mod.rs [hunk 4]
```diff
@@ -153,11 +166,33 @@ impl<T: BlockchainCtx> BlockchainDataBuilder<T> {
     }
 }
 
+#[derive(Clone, Debug)]
+pub struct Statistic {
```
Comment (sokyrko) [inline mamoru-core/src/blockchain_data/mod.rs:170]: Feel free to implement `Copy` for this struct.

File: mamoru-sniffer-c/src/evm_types.rs [hunk 5]
```diff
@@ -50,6 +50,19 @@ fn evm_blockchain_data_builder_set_mempool_source(builder: &mut FfiEvmBlockchain
     builder.inner.set_mempool_source();
 }
 
+#[ffi_export]
+fn evm_blockchain_data_builder_set_statistic(
+    builder: &mut FfiEvmBlockchainDataBuilder,
```
Comment (sokyrko) [inline mamoru-sniffer-c/src/evm_types.rs:55]: `add_statistics`

File: mamoru-sniffer/src/sniffer.rs [hunk 6]
```diff
@@ -115,6 +133,11 @@ impl Sniffer {
         fields(tx = ?ctx.tx(), block = ?ctx.block(), source = ?ctx.source(), level = "debug")
     )]
     pub async fn observe_data<T: BlockchainCtx>(&self, ctx: BlockchainData<T>) {
+        self.statistic_tx
+            .send(ctx.statistic().clone())
+            .await
+            .expect("Failed to send statistic");
```
Comment (sokyrko) [inline mamoru-sniffer/src/sniffer.rs:139]: Handle the failure case properly, see `report_tx` below for examples.

File: mamoru-sniffer/src/sniffer.rs [hunk 7]
```diff
@@ -115,6 +133,11 @@ impl Sniffer {
         fields(tx = ?ctx.tx(), block = ?ctx.block(), source = ?ctx.source(), level = "debug")
     )]
     pub async fn observe_data<T: BlockchainCtx>(&self, ctx: BlockchainData<T>) {
+        self.statistic_tx
+            .send(ctx.statistic().clone())
```
Comment (sokyrko) [inline mamoru-sniffer/src/sniffer.rs:137]: No need to `clone` after you implement `Copy`.

File: mamoru-sniffer/src/sniffer.rs [hunk 8]
```diff
@@ -193,6 +219,9 @@ impl SnifferBgTask {
         daemons_update_interval: Duration,
         incident_send_interval: Duration,
         max_incident_batch_size: usize,
+        statistic_send_interval: Duration,
+        statistic_rx: Receiver<Statistic>,
+        statistic_buffer_size: usize,
```
Comment (sokyrko) [inline mamoru-sniffer/src/sniffer.rs:224]: It's time to refactor BgTask. Make `BgTaskConfig` and put all configuration stuff there:
```
struct BgTaskConfig {
        daemons_update_interval: Duration,
        incident_send_interval: Duration,
        max_incident_batch_size: usize,
        statistic_send_interval: Duration,
        statistic_rx: Receiver<Statistic>,
        statistic_buffer_size: usize,
}
```

File: mamoru-sniffer/src/sniffer.rs [hunk 9]
```diff
@@ -269,14 +306,32 @@ impl SnifferBgTask {
...
+                }
             }
         }
     }
 
     /// Receives incidents from the channel up to `max_incident_batch_size`.
     /// Returns `TryRecvError::Disconnected` if the channel is closed.
     fn receive_incidents(&mut self) -> Result<Vec<IncidentReport>, TryRecvError> {
-        let mut items = Vec::with_capacity(self.max_incident_batch_size);
+        let mut items = Vec::with_capacity(self.statistic_buffer_size);
```
Comment (sokyrko) [inline mamoru-sniffer/src/sniffer.rs:334]: This is a mistake, keep `max_incident_batch_size` here.

File: mamoru-sniffer/src/sniffer.rs [hunk 10]
```diff
@@ -269,14 +306,32 @@ impl SnifferBgTask {
                         }
                     }
                 }
+
+                _ = statistic_interval.tick() => {
+                    match self.receive_statistic() {
```
Comment (sokyrko) [inline mamoru-sniffer/src/sniffer.rs:311]: Sending statistics may interfere with incident reporting... We need to make it async, as we send statistics often.

File: mamoru-sniffer/src/validation_chain/message_client.rs [hunk 11]
```diff
@@ -388,6 +391,39 @@ impl MessageClient {
...
+        &self,
+        statistics: impl IntoIterator<Item = Statistic>,
+    ) -> ClientResult<Vec<MsgMarkSnifferStatisticResponse>> {
+        let sniffer = self.config.address().to_string();
+
+        let report_messages: Vec<_> = statistics
+            .into_iter()
+            .map(|stat: Statistic| MsgMarkSnifferStatistic {
+                creator: sniffer.clone(),
+                statistic: Option::from(SnifferStatistic {
```
Comment (sokyrko) [inline mamoru-sniffer/src/validation_chain/message_client.rs:405]: Some(SnifferStatistics {...})

**Comment by sokyrko** [comment]
@gofmanaa I don't see a way to trigger sending statistics from a sniffer. Please add it.

File: mamoru-core/src/blockchain_data/mod.rs (hunk 3 above)
Comment (gofmanaa) [reply mamoru-core/src/blockchain_data/mod.rs]: fixed

File: mamoru-core/src/blockchain_data/mod.rs (hunk 4 above)
Comment (gofmanaa) [reply mamoru-core/src/blockchain_data/mod.rs]: fixed

File: mamoru-sniffer/src/sniffer.rs (hunk 7 above)
Comment (gofmanaa) [reply mamoru-sniffer/src/sniffer.rs]: fixed

File: mamoru-sniffer/src/sniffer.rs (hunk 8 above)
Comment (gofmanaa) [reply mamoru-sniffer/src/sniffer.rs]: implemented

File: mamoru-sniffer/src/sniffer.rs (hunk 9 above)
Comment (gofmanaa) [reply mamoru-sniffer/src/sniffer.rs]: fixed

File: mamoru-sniffer/src/sniffer.rs (hunk 10 above)
Comment (gofmanaa) [reply mamoru-sniffer/src/sniffer.rs]: fixed

File: mamoru-sniffer/src/validation_chain/message_client.rs (hunk 11 above)
Comment (gofmanaa) [reply mamoru-sniffer/src/validation_chain/message_client.rs]: fixed

File: mamoru-core/src/blockchain_data/mod.rs (hunk 1 above)
Comment (gofmanaa) [reply mamoru-core/src/blockchain_data/mod.rs]: fixed

File: mamoru-core/src/blockchain_data/mod.rs [hunk 12]
```diff
@@ -74,6 +74,7 @@ pub struct BlockchainDataBuilder<T> {
     source: DataSource,
     tx: Option<(Id, Hash)>,
     block: Option<(Id, Hash)>,
+    statistic: Option<Statistic>,
```
[the remark this answers is not in the document]
Comment (gofmanaa) [reply mamoru-core/src/blockchain_data/mod.rs]: fixed

File: mamoru-sniffer/src/sniffer.rs [hunk 13]
```diff
@@ -190,9 +256,7 @@ impl SnifferBgTask {
         daemons: Arc<RwLock<Vec<Daemon>>>,
```
Comment (sokyrko) [inline mamoru-sniffer/src/sniffer.rs:256]: You can remove `#[allow(clippy::too_many_arguments)]` now

File: mamoru-sniffer/src/statistics_bg_task.rs [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+                    debug!(?statistics, "Reporting statistics...");
+                    if let Err(err) = self
+                        .message_client
+                        .mark_sniffer_statistic(vec![statistics])
+                        .await
+                    {
+                        error!(error = ?err, "Failed to report statistic")
+                    }
+                }
+                None => continue,
```
Comment (sokyrko) [inline mamoru-sniffer/src/statistics_bg_task.rs:67]: `None` here probably means that the channel is closed.
You may write this block like
```
while let Some(statistics) = self.statistic_rx.recv().await {
...
}
``

File: mamoru-sniffer/src/statistics_bg_task.rs [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+                }
+            }
+        }
+    }
+
+    /// Receive statistic from sniffer and send it to Validation Chain.
+    async fn receive_statistic(&mut self) -> Result<Vec<StatisticsReport>, TryRecvError> {
+        let mut items = Vec::with_capacity(self.task_config.buffer_size);
+        loop {
+            match self.statistic_rx.recv().await {
```
Comment (sokyrko) [inline mamoru-sniffer/src/statistics_bg_task.rs:106]: `recv` blocks until you recenve `Some(T)`.

File: mamoru-sniffer/src/statistics_bg_task.rs [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+        None
+    }
+    pub fn default_buffer_size() -> usize {
+        256
+    }
+}
+
+pub struct StatisticBgTask {
+    message_client: MessageClient,
+    pub statistic_rx: Receiver<StatisticsReport>,
```
Comment (sokyrko) [inline mamoru-sniffer/src/statistics_bg_task.rs:30]: Does it need to be `pub`?

File: mamoru-sniffer/src/statistics_bg_task.rs [hunk 17]
```diff
@@ -0,0 +1,121 @@
...
+    pub async fn run(self) {
+        match self.task_config.send_interval_sec {
+            None => self.run_immediately().await,
+            Some(_) => self.run_interval().await,
+        };
+    }
+
+    async fn run_immediately(mut self) {
+        loop {
+            match self.statistic_rx.recv().await {
```
Comment (sokyrko) [inline mamoru-sniffer/src/statistics_bg_task.rs:56]: I think it's better to take an array here as well, as you may receive statistics faster than you send it.

**Review by sokyrko — APPROVED** [review]

**Pull request merged**